[FEAT] 3주차 과제

mminny opened the pull request:
# 🔥Pull requests

## 👷작업한 내용
<!-- 작업한 내용을 적어주세요. -->
### 🌱기본과제
* Compositional layout을 이용한 UI구현
원래는 TableView와 CollectionView를 중첩해서 구현할 생각이었는데 그러면 다음주에 PR을 올릴 것 같아서 Compositional layout으로 구현을 해봤어요. 오히려 이 레이아웃으로 구현하니까 구조를 더 생각하기 쉬웠던 것 같아요. 그래도... 아직 했갈리는 것들이 많아서 CollectionView를 중첩시켜서 해보긴 해야할 것 같네요... 레이아웃 수정도 추가하겠습니다

### 🧨심화과제
* 각 Cell 마다 스크롤 구현(시도했으나 실패...)
다른 부분은 시간 이슈로 구현을 못해서 너무 아쉽쓰... 계속 추가해서 완성해볼게요!

## 💭되돌아보기
Cell을 등록해서 하는 것도 Delegate해서 데이터 전달하는 것도 알겠는데 Cell하고 View 다 만들고 뷰컨에 연결하는 과정에서 이해가 부족하지 않았나 생각해요... 음... compositional layout도 이해를 좀 더 해봐야 할 것 같습니닷 좀 반복해서 연습해봐야지... 

## 📸스크린샷
|기능|스크린샷|GIF|
|:--:|:--:|:--:|
|Compositional layout을 이용한 UI구현|none|none| 

## 📟관련 이슈
- Resolved: #4

mminny committed:
[Fix] fix error

mminny committed:
[Fix] LogoView 수정

seuriseuljjeok commented on Week2-Assignment-Tving/Sopt-Assignment/Cells/PopularLiveChannelCollectionViewCell.swift:
```diff
@@ -0,0 +1,103 @@
+    private func setLayout() {
+        liveChannelImage.snp.makeConstraints {
+            $0.top.leading.trailing.equalToSuperview()
+            $0.width.equalTo(160)
```
leading과 trailing을 잡아주면 width는 굳이 안잡아줘도 될 것 같아요!

seuriseuljjeok commented on Week2-Assignment-Tving/Sopt-Assignment/Cells/RecommandCollectionViewCell.swift:
```diff
@@ -0,0 +1,72 @@
+    
+    private func setLayout() {
+        vstackView.snp.makeConstraints {
+            $0.top.equalToSuperview().offset(583)
```
지금 스택뷰의 레이아웃은 전체 뷰를 기준으로 잡아준 것 같은데 UICollectionViewCompositionalLayout로 셀의 레이아웃을 잡아줬기 때문에 셀을 기준으로 레이아웃을 잡아줘야 할 것 같아요!
이 방법으로 레이아웃을 잡았을 때 혹시 잘 잡혔나요?

seuriseuljjeok commented on Week2-Assignment-Tving/Sopt-Assignment/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,136 @@
+    
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        if section == 0 || section == 1 {
+            return 8
```
딱 숫자로 지정하기보다는 데이터를 담아둔 배열의 크기로 지정하는 것이 더 좋을 것 같아요!

seuriseuljjeok reviewed (approved):
이번 과제 많이 어려웠을텐데 과제 하시느라 수고하셨습니다아🍏

JinUng41 commented on Week2-Assignment-Tving/Sopt-Assignment/Cells/MainCollectionViewCell.swift:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+// MARK: Extensions
+extension MainCollectionViewCell {
+    func setHierarchy() {
+        addSubview(mainCellImageView)
+        mainCellImageView.addSubviews(VStackView)
+    }
+    
+    func setLayout() {
+        mainCellImageView.snp.makeConstraints { $0.edges.equalToSuperview() }
+        
+        VStackView.snp.makeConstraints {
+            $0.leading.equalTo(mainCellImageView.snp.leading).offset(16)
+            $0.bottom.equalTo(mainCellImageView.snp.bottom).inset(28)
+        }
+
+    }
```
두 메서드는 셀의 오토레이아웃을 설정하기 위한 코드로 보여요. 셀 외부에서 호출하기 위한 함수가 아니기에 `private`으로 접근 제어를 두어, 내용을 은닉화 해보는 건 어떨까요?

JinUng41 commented on Week2-Assignment-Tving/Sopt-Assignment/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        setRegister()
+    }
+    
+    func setDelegate() {
+        homeView.collectionView.delegate = self
+        homeView.collectionView.dataSource = self
+    }
+    
+    func setRegister() {
+        // header cell 등록
+        homeView.collectionView.register(HomeViewHeaderCell.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HomeViewHeaderCell")
+        homeView.collectionView.register(MainCollectionViewCell.self, forCellWithReuseIdentifier: HomeSection.main.identifier)
+        homeView.collectionView.register(RecommandCollectionViewCell.self, forCellWithReuseIdentifier: HomeSection.recommand.identifier)
+        homeView.collectionView.register(PopularLiveChannelCollectionViewCell.self, forCellWithReuseIdentifier: HomeSection.live.identifier)
+    }
```
오, Delegate와 더불어 셀 등록 코드를 묶어 좀 더 해당 내용에 집중할 수 있는 것 같아요. 저도 참고해 보겠습니다.
다만, 'header cell 등록' 주석은 이미 _HomeViewHeaderCell_ 을 등록하고 있는 곳이 타입의 이름에서부터 드러나기 때문에, 필요없어도 괜찮지 않을까 하는 생각입니다!

JinUng41 commented on Week2-Assignment-Tving/Sopt-Assignment/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,136 @@
+        }
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        let section = HomeSection(rawValue: indexPath.section)!
+        
+        switch section {
+        case .main:
+            let cell = collectionView.dequeueReusableCell(withReuseIdentifier: HomeSection.main.identifier, for: indexPath) as! MainCollectionViewCell
```
_Enum_ 을 활용하여 섹션에 대한 분기처리를 시도하신 것이 인상적입니다!. 다만 강제 언래핑은 위험해 보이는데요. 강제 언래핑을 하신 이유를 여쭤보고 싶어요.

JinUng41 commented on Week2-Assignment-Tving/Sopt-Assignment/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,136 @@
+            return cell
+        }
+    }
+    
+    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
+        
+        guard let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: HomeViewHeaderCell.className, for: indexPath) as? HomeViewHeaderCell else {
+            return UICollectionReusableView()
+        }
```
이렇게 번거롭더라도 옵셔널 바인딩을 이용하여 값을 추출하는 코드가 훨씬 안전하다고 생각합니다! 👍

JinUng41 reviewed (approved):
민성님, 과제 수행하시느라 정말 고생많으셨습니다~ 👏👏
몇 가지 간단한 코멘트 남겨두었으니 확인해 주시면 감사하겠습니당~

nolanMinsung commented on Week2-Assignment-Tving/Sopt-Assignment/Cells/MainCollectionViewCell.swift:
```diff
@@ -0,0 +1,76 @@
+// MARK: Extensions
+extension MainCollectionViewCell {
+    func setHierarchy() {
+        addSubview(mainCellImageView)
```
애플 공식문서에서는 UICollectoinView에 뷰를 subView로 넣을 경우, cell의 subView로 직접 넣지 말고,
UICollectionViewCell의 contentView라는 속성의 subView로 넣으라고 되어 있습니다! 😁
혹시라도 의도적으로 cell의 addSubView를 직접 호출하여 구현하신 거라면 왜 그렇게 구현하셨는지 궁금합니다~

"Alternatively, add the views needed to present the data item’s content as subviews to the view in the **contentView** property. Don’t directly add subviews to the cell itself."

그러니 아마 해당 코드는
self.contentView.addSubview(posterImageView)
로 바꾸는 것이 적절해 보입니다!~👍

관련하여 애플 공식 문서 첨부해 드립니다!
https://developer.apple.com/documentation/uikit/uicollectionviewcell

nolanMinsung commented on Week2-Assignment-Tving/Sopt-Assignment/View/HomeView.swift:
```diff
@@ -0,0 +1,194 @@
+    
+    // MARK: CollectionView Layout
+    
+    private func makeCollectionViewLayout() -> UICollectionViewCompositionalLayout {
```
compositional Layout 구현하시느라 수고 많으셨습니다...!!👍👍

nolanMinsung reviewed (approved):
이번 과제도 정말 수고 많으셨습니다!!😁
앱잼까지 파이팅!!🔥🔥🔥

mminny committed:
commit for testing program

mminny committed:
[Fix] 상단 로고 뷰 레이아웃 수정 및 네비게이션 바 숨김

mminny committed:
[Fix] 이미지 asset 추가 및 RecommandCollectionViewCell의 subview가 뜨지 않는 문제 해결

문제원인: UIStackView에 subview를 arrange해놓긴 했는데 setHierarchy() 함수에서 addSubviews()에 UIStackView만 등록해놓은 것이 문제였던 것 같음.
궁금증: 왜 이미 UIStackView에 subview들이 정렬되어있는데  VStackView만 등록하면 그 안에 있는 subview들도 묶음 상품처럼 계층적 구조가 생기지 않는 이유가 무엇인지 궁금함. 왜 따로 또 addSubviews에 넣어줘야 하는지 궁금하다~

mminny committed:
[Docs] 필요없는 주석 삭제 및 변수 이름 수정과 이미지 resource 이름 변경

mminny committed:
[Docs] 클래스 이름 날아간거 수정

mminny committed:
commit for test
